File: bindings/java/src/main/java/ai/arcships/aimux/HostOperation.java
package ai.arcships.aimux;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.sun.jna.Pointer;
import com.sun.jna.ptr.IntByReference;
import com.sun.jna.ptr.PointerByReference;
import java.io.Closeable;
import java.io.IOException;
import java.util.Spliterator;
import java.util.Spliterators;
import java.util.concurrent.CancellationException;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Consumer;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

/** Data-only C transport. No Java object is retained by Rust. */
final class HostOperation implements Closeable {
    private final AtomicLong handle;
    private final ToolCallRepairFunction repair;

    HostOperation(Model model, String mode, String prompt, String options, ToolCallRepairFunction repair) {
        ObjectNode request = Types.AimuxJson.MAPPER.createObjectNode();
        request.put("protocol_version", 1).put("mode", mode);
        request.set("prompt", parse(prompt));
        request.set("options", parse(options == null ? "{}" : options));
        request.put("repair_tool_call", true);
        this.repair = repair;
        handle = new AtomicLong(model.startOperation(request.toString()));
    }

    private static JsonNode parse(String text) {
        try { return Types.AimuxJson.MAPPER.readTree(text); }
        catch (IOException e) { throw new IllegalArgumentException("invalid operation JSON", e); }
    }

    /** ANY coordinator preserves the caller's thread for synchronous hooks. */
    String next() {
        while (handle.get() != 0) {
            if (Thread.currentThread().isInterrupted()) { close(); throw new CancellationException("operation interrupted"); }
            long h = handle.get();
            PointerByReference out = new PointerByReference();
            IntByReference state = new IntByReference();
            Pointer error = AimuxFFI.INSTANCE.aimux_operation_next(h, 2, 100, out, state);
            if (error != null) {
                RuntimeException failure = AimuxResult.expectAimuxError(error, "operation_next");
                if (handle.get() == 0) return null;
                throw failure;
            }
            if (state.getValue() == 1) continue;
            if (state.getValue() == 2) return null;
            if (state.getValue() != 0) throw new IllegalStateException("concurrent operation reader");
            String wire;
            try { wire = out.getValue().getString(0, "UTF-8"); }
            finally { AimuxFFI.INSTANCE.aimux_free_string(out.getValue()); }
            JsonNode event = parse(wire);
            String type = event.path("type").asText();
            if (!type.equals("repair_request")) return event.get(type.equals("part") ? "part" : "result").toString();
            ObjectNode reply = Types.AimuxJson.MAPPER.createObjectNode();
            try {
                Types.ToolCallRepairContext context = Types.AimuxJson.MAPPER.treeToValue(event.get("context"), Types.ToolCallRepairContext.class);
                Types.RawToolCall value = repair.repair(context);
                if (value == null) reply.put("type", "unchanged");
                else {
                    if (value.getToolCallId() == null || value.getToolName() == null || value.getInput() == null) throw new IllegalArgumentException("repair returned null required fields");
                    reply.put("type", "repaired").set("tool_call", Types.AimuxJson.MAPPER.valueToTree(value));
                }
            } catch (Exception e) { reply.removeAll(); reply.put("type", "failed").put("message", e.toString()); }
            if (handle.get() == 0) return null;
            IntByReference status = new IntByReference();
            error = AimuxFFI.INSTANCE.aimux_operation_reply(h, event.path("request_id").asText(), reply.toString(), status);
            if (error != null) throw AimuxResult.expectAimuxError(error, "operation_reply");
            if (status.getValue() != 0 && status.getValue() != 3) throw new IllegalStateException("unexpected operation reply status");
        }
        return null;
    }

    @Override public void close() {
        long h = handle.getAndSet(0);
        if (h != 0) AimuxFFI.INSTANCE.aimux_operation_drop(h);
    }

    static String result(Model model, String mode, String prompt, String options, ToolCallRepairFunction repair) {
        try (HostOperation op = new HostOperation(model, mode, prompt, options, repair)) {
            String value = op.next();
            if (value == null) throw new IllegalStateException("operation ended without a result");
            return value;
        }
    }

    /** Callers must close a short-circuited stream (try-with-resources). */
    static Stream<String> stream(Model model, String mode, String prompt, String options, ToolCallRepairFunction repair) {
        HostOperation op = new HostOperation(model, mode, prompt, options, repair);
        return StreamSupport.stream(new Spliterators.AbstractSpliterator<String>(Long.MAX_VALUE, Spliterator.ORDERED) {
            @Override public boolean tryAdvance(Consumer<? super String> consumer) {
                try {
                    String value = op.next();
                    if (value == null) { op.close(); return false; }
                    consumer.accept(value); return true;
                } catch (RuntimeException | Error e) { op.close(); throw e; }
            }
        }, false).onClose(op::close);
    }
}

File: bindings/kotlin/src/main/kotlin/ai/arcships/aimux/HostOperation.kt
package ai.arcships.aimux

import com.sun.jna.ptr.IntByReference
import com.sun.jna.ptr.PointerByReference
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import java.io.Closeable
import java.util.concurrent.CancellationException
import java.util.concurrent.atomic.AtomicLong

/** Pure data transport; user functions run after each native call returns. */
internal class HostOperation(model: Model, mode: String, prompt: String, options: String?, private val repair: ToolCallRepair) : Closeable {
    private val handle = AtomicLong(model.startOperation(buildJsonObject {
        put("protocol_version", 1); put("mode", mode)
        put("prompt", AimuxJson.parseToJsonElement(prompt))
        put("options", AimuxJson.parseToJsonElement(options ?: "{}"))
        put("repair_tool_call", true)
    }.toString()))

    fun next(): String? {
        while (handle.get() != 0L) {
            if (Thread.currentThread().isInterrupted) { close(); throw CancellationException("operation interrupted") }
            val h = handle.get()
            val out = PointerByReference(); val state = IntByReference()
            FFI.lib.aimux_operation_next(h, 2, 100, out, state)?.let { throw expectAimuxError(it) }
            if (state.value == 1) continue
            if (state.value == 2) return null
            check(state.value == 0) { "concurrent operation reader" }
            val event = AimuxJson.parseToJsonElement(takeString(out.value)!!).jsonObject
            val type = event.getValue("type").jsonPrimitive.content
            if (type != "repair_request") return event.getValue(if (type == "part") "part" else "result").toString()
            val reply = try {
                val context = AimuxJson.decodeFromJsonElement(ToolCallRepairContext.serializer(), event.getValue("context"))
                val value = repair(context)
                buildJsonObject {
                    put("type", if (value == null) "unchanged" else "repaired")
                    if (value != null) put("tool_call", AimuxJson.encodeToJsonElement(RawToolCall.serializer(), value))
                }
            } catch (e: CancellationException) { throw e }
            catch (e: InterruptedException) { Thread.currentThread().interrupt(); throw CancellationException("repair interrupted") }
            catch (e: Exception) { buildJsonObject { put("type", "failed"); put("message", e.toString()) } }
            if (handle.get() == 0L) return null
            val status = IntByReference()
            FFI.lib.aimux_operation_reply(h, event.getValue("request_id").jsonPrimitive.content, reply.toString(), status)?.let { throw expectAimuxError(it) }
            check(status.value == 0 || status.value == 3) { "unexpected operation reply status" }
        }
        return null
    }

    override fun close() { handle.getAndSet(0).takeIf { it != 0L }?.let { FFI.lib.aimux_operation_drop(it) } }

    companion object {
        fun result(model: Model, mode: String, prompt: String, options: String?, repair: ToolCallRepair): String =
            HostOperation(model, mode, prompt, options, repair).use { it.next() ?: error("operation ended without result") }

        fun stream(model: Model, mode: String, prompt: String, options: String?, repair: ToolCallRepair, emit: (String) -> Unit) {
            HostOperation(model, mode, prompt, options, repair).use { op ->
                while (true) emit(op.next() ?: break)
            }
        }
    }
}
